src/xinterpreter.cpp

```diff
@@ -111,14 +119,33 @@ SEXP try_parse(const std::string& code, int execution_counter) {
             return xeus::create_error_reply();
         }
 
-        // TODO: wrap in a tryCatch
-        // TODO: eval not just the first, but all
-        SEXP out = PROTECT(Rf_eval(VECTOR_ELT(parsed, 0), R_GlobalEnv));
-        
-        // echo the code for now
-        nl::json pub_data;
-        pub_data["text/plain"] = "bonjour"; // REAL(out)[0];
-        // publish_execution_result(execution_counter, std::move(pub_data), nl::json::object());
+        R_xlen_t n = XLENGTH(parsed);
+        int ErrorOccurred;
+        SEXP smb_withVisible = Rf_install("withVisible");
+            
+        for (R_xlen_t i = 0; i < n; i++) {
+            SEXP expr   = PROTECT(Rf_lang2(smb_withVisible, VECTOR_ELT(parsed, i)));
+            SEXP result = PROTECT(R_tryEval(expr, R_GlobalEnv, &ErrorOccurred));
+            bool visible = LOGICAL(VECTOR_ELT(result, 1))[0];
+
+            if (!ErrorOccurred && visible) {
+                capture_stream.str("");
+                ptr_R_WriteConsoleEx = capture_WriteConsoleEx;
+                SEXP value = VECTOR_ELT(result, 0);
+
+                R_ToplevelExec([](void* value) {
+                    Rf_PrintValue((SEXP)value);
+                }, (void*)value);
+                
+                ptr_R_WriteConsoleEx = WriteConsoleEx;
+                
+                nl::json pub_data;
+                pub_data["text/plain"] = capture_stream.str();
+                publish_execution_result(execution_counter, std::move(pub_data), nl::json::object());
```

Comment:
I wonder if this should go outside of the for-loop. The for-loop would concatenate all captured streams, separated with new lines, and `publish_execution_result` would be called just once.

This would result in having only one output number (`[1]`), instead of many. This is the behavior I'd expect from the kernel, though I guess it's not a kernel protocol requirement.

![Screenshot from 2023-10-09 08-43-51](https://github.com/jupyter-xeus/xeus-r/assets/21197331/78741a6b-da54-4061-b1a5-244356e3c938)

Reply:
That makes sense. Follow up question. This is based on what happens at the R console when multiple expressions are used: 

<img width="104" alt="image" src="https://github.com/jupyter-xeus/xeus-r/assets/2625526/29561fd0-1953-46cb-8f98-092b0091d758">

Bu the 🐍 kernel seems to behave differently: 

<img width="203" alt="image" src="https://github.com/jupyter-xeus/xeus-r/assets/2625526/57a87c2c-0c2a-45f1-93d5-0bab083352d8">

[irkernel] does something else: 

<img width="239" alt="image" src="https://github.com/jupyter-xeus/xeus-r/assets/2625526/7caa1c4e-0b4f-4990-bc4f-cb0ba71fcda0">

So I'm confused. Should this be modelled on the 🐍 behaviour, i.e. only `publish_execution_result()` the last result ?

Reply:
Ah sorry, I read a bit too fast your example and got confused. 

The kernel should only show the last statement indeed, using `publish_execution_result` only once.

Reply:
Though if you think users would expect it to behave like the R console, then maybe that's the way to go.

Reply:
I think it should display "the result of the cell", as what the 🐍  does and IIUC what the `xeus-cling` kernel also does. I'll follow up in another PR.